### ISSUE TITLE
default value given in configuration was set correctly in extension but not assigned in compiler pass

### ISSUE THREAD

**Issue by jerryladoo** [symfony/symfony#19149]
I have created the following configs

``` yaml
// config.yaml
export:
    archiver:
        zip:
            target: destination.local
            destination: local
            working_dir: ~
            use: ~
            lib:
                command:
                  standard: 'zip -r %%s %%s'
```

and defined them in configuration with NodeBuilder

``` php
// configuration.php
/**
  * {@inheritdoc}
  */
public function getConfigTreeBuilder()
{
        $treeBuilder = new TreeBuilder();
        $rootNode = $treeBuilder->root('export');

        // Here you should define the parameters that are allowed to
        // configure your bundle. See the documentation linked above for
        // more information on that topic.

        $this->addArchiverNode($rootNode);

        return $treeBuilder;
}

private function addArchiverNode(ArrayNodeDefinition $rootNode)
{
        // create all fields
        $fields = [
            ["name" => "target", "type" => "scalar", "isRequired" => true, "cannotBeEmpty" => true],
            ["name" => "destination", "type" => "scalar", "default" => "local"],
            ["name" => "working_dir", "type" => "scalar", "default" => "\tmp"],
            ["name" => "use", "type" => "scalar", "default" => "ZipArchive"],
            ["name" => "standard", "type" => "scalar"]
        ];
        $nodes = $this->createNodes($fields, false, false);

        /**
         * build command field
         * command:
         *      standard: ~
         **/
        $commandNode = $this->createArrayNode("command", [$nodes["standard"]]);

        /**
         * more could be added in for lib node
         *
         * build lib field
         * lib:
         *      command:
         *          <...>
         **/
        $libNode = $this->createArrayNode("lib", [$commandNode]);

        /**
         * more could be added in for lib node
         *
         * build archiver field
         * archiver:
         *      target: ~
         *      destination: ~
         *      working_dir:
         *      use: ~
         *      lib:
         *          <...>
         **/
        $archiverNode = $this->createArrayNode("archiver", [$nodes["target"], $nodes["destination"], $nodes["working_dir"], $nodes["use"], $libNode], [
            "prototype" => "array", "requiresAtLeastOneElement" => true, "useAttributeAsKey" => "id", "addDefaultsIfNotSet" => true
        ]);
        $rootNode->append($archiverNode);
}

private function createNodes($nodes, $returnNodeIfSingle = false, $returnOnlyValues = true)
{
        $nodeBuilder = new NodeBuilder();
        $list = [];
        foreach($nodes as $node) {
            $newNode = $nodeBuilder->node($node["name"], $node["type"]);
            if (array_key_exists("default", $node)) {
                $defaultValue = $node["default"];
                $newNode->defaultValue($defaultValue);  // <- this works in extension but does not work in compiler pass
                $newNode->validate()->ifNull()->then(function($v) use($defaultValue) {return $defaultValue;}); // if '~' is given,then it will be reset with default value, but same issue as above occurs
            }

            if (array_key_exists("isRequired", $node) && $node["isRequired"]) $newNode->isRequired();
            if (array_key_exists("cannotBeEmpty", $node) && $node["cannotBeEmpty"]) $newNode->cannotBeEmpty();

            $list[$node["name"]] = $newNode;
        }
        if (count($list) == 1 && $returnNodeIfSingle) return array_shift($list);
        $values = array_values($list);

        return $returnOnlyValues ? $values : $list;
}

private function createArrayNode($name, $children = [], $options = [])
{
        $builder = new NodeBuilder();
        $arrayNode = $builder->arrayNode($name);
        if (empty($children) && empty($options)) return $arrayNode;

        // append all children to normal array node
        if (!empty($children) && empty($options)) {
            foreach($children as $child) {
                $arrayNode->append($child);
            }
            return $arrayNode;
        }

        if (isset($options["requiresAtLeastOneElement"])) $arrayNode->requiresAtLeastOneElement();
        if (isset($options["useAttributeAsKey"])) $arrayNode->useAttributeAsKey($options["useAttributeAsKey"]);

        // create prototype for $arrayNode
        if (isset($options["prototype"])) {
            $node = $arrayNode->prototype($options["prototype"]);
            if (!$node instanceof ArrayNodeDefinition) throw new \InvalidArgumentException("currently only support array prototype.");
            if ($node instanceof ArrayNodeDefinition && empty($children)) throw new \InvalidArgumentException('$children is required when creating an array prototype.');
            if (isset($options["addDefaultsIfNotSet"])) $node->addDefaultsIfNotSet(); // allow adding defaults

            foreach($children as $child) {
                $node->append($child);
            }
        } else {
            if (isset($options["addDefaultsIfNotSet"])) $arrayNode->addDefaultsIfNotSet();
        }

        return $arrayNode;
}
```

and then I ran and printed out the config in extension and compiler pass respectively

``` php
// ExportExtension.php
class ExportExtension extends Extension
{
    /**
     * {@inheritdoc}
     */
    public function load(array $configs, ContainerBuilder $container)
    {
        $configuration = new Configuration();
        $config = $this->processConfiguration($configuration, $configs);

        $loader = new Loader\YamlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
        $loader->load('services.yml');

        print_r(json_encode($config));print_r(PHP_EOL);
    }
}

// the output:
{
  "archiver": {
    "zip": {
      "target": "destination.local",
      "destination": "local",
      "working_dir": "/tmp",  // <- default value has been set
      "use": "ZipArchive",    // <- default value has been set
      "lib": {
        "command": {
          "standard": "zip -r %%s %%s"
        }
      }
    }
  }
}
```

in extension it outputted correct config settings, but in compiler pass below

``` php
class TaggedServiceCompilerPass implements CompilerPassInterface
{
    /**
     * You can modify the container here before it is dumped to PHP code.
     *
     * @param ContainerBuilder $container
     */
    public function process(ContainerBuilder $container)
    {
        $taggedServiceManagerId = "export_manager";
        $tagName = "export.archiver";
        if (empty($taggedServiceManagerId) || !$container->has($taggedServiceManagerId) || empty($tagName)) {
            return;
        }
        $definition = $container->findDefinition($taggedServiceManagerId);
        $taggedServices = $container->findTaggedServiceIds($tagName);

        // get the actual config settings
        $extensionConfig = $container->getExtensionConfig("export");
        $extensionConfig = array_shift($extensionConfig);
print_r(json_encode($extensionConfig));print_r(PHP_EOL);

        foreach ($taggedServices as $id => $tags) {
            foreach ($tags as $attributes) {
                $this->addMethodCall($definition, new Reference($id), $attributes);
            }
        }
    }
}

// the output:
{
  "archiver": {
    "zip": {
      "target": "destination.local",
      "destination": "local",
      "working_dir": null,  // <- it has been set to null but not default value
      "use": null,              // <- same as above
      "lib": {
        "command": {
          "standard": "zip -r %%s %%s"
        }
      }
    }
  }
}
```

``` php
// ExportBundle.php
class ExportBundle extends Bundle
{
    public function build(ContainerBuilder $container)
    {
        parent::build($container);
        $container->addCompilerPass(new ArchiverCompilerPass());
    }
}
```

it outputted an incorrect settings. All default values were not set properly. The execution order I found was
1. ExportExtension.php
2. TaggedServiceCompilerPass.php

Even if I remove `working_dir` and `use` from config as below

``` yaml
export:
    archiver:
        zip:
            target: destination.local
            destination: local
            lib:
                command:
                  standard: 'zip -r %%s %%s'
```

in extension it still outputted correct settings as below:

``` php
// the output
{
  "archiver": {
    "zip": {
      "target": "destination.local",
      "destination": "local",
      "lib": {
        "command": {
          "standard": "zip -r %%s %%s"
        }
      },
      "working_dir": "/tmp",  // <-- this field has been added in and with default value as expected
      "use": "ZipArchive"    // <-- same as above
    }
  }
}
```

in compiler pass on the other hand it still outputted incorrect settings

``` php
// TaggedServiceCompilerPass.php
// output
{
  "archiver": {
    "zip": {
      "target": "destination.local",
      "destination": "local",
      "lib": {
        "command": {
          "standard": "zip -r %%s %%s"
        }
      }
    }
  }
}
```

Any ideas guys? Did I miss anything?

**Comment by jerryladoo** [comment]
I forgot to mention that the version I am using now is 2.7.x

**Comment by xabbuh** [comment]
@jerryladoo Would you mind forking the Symfony Standard Edition and make the changes that are necessary to reproduce your issue?

**Comment by jerryladoo** [comment]
@xabbuh how to do that inside symfony itself?

**Comment by xabbuh** [comment]
@jerryladoo Sorry, I do not understand your question. What I meant is that we would need to be able to reproduce the behaviour you observe. So if you could clone the Symfony Standard Edition and make changes to it on your clone that lead to the same behaviour, we would be able to look into your issue.

**Comment by jerryladoo** [comment]
@xabbuh To be honest, I am not sure which part of symfony I should lay down these changes.
what I did was
1. create a new project
2. install symfony 2.7 in the project 
3. create all those classes mentioned above in the new project
4. run any commands that can trigger dependecyinjection

and then it happened.

**Comment by xabbuh** [comment]
@jerryladoo Could you upload that project somewhere so that we can reproduce your issue?

**Comment by javiereguiluz** [comment]
@jerryladoo as you can see, at the moment we have more than 600 pending issues in this repository. That's why we ask issue reporters to help us preparing the code needed to reproduce this issue. Specifically, we'd need you to:

1. Fork the official Symfony repository
2. Create a new branch called "bug_19149"
3. Add the code needed to reproduce your issue.
4. Push this branch to your fork
5. Tell us the URL of your fork so we can fetch your branch and reproduce the issue.

Thanks!

**Comment by Simperfit** [comment]
@javiereguiluz Did you mean the [Symfony Standard Edition](https://github.com/symfony/symfony-standard/tree/2.7) ?

@jerryladoo what @xabbuh wanted you to do is to
- Fork the [Symfony Standard Edition](https://github.com/symfony/symfony-standard)
- Checkout the 2.7 branch
- do all the modification that leads to this issue and then post the link in here, for the symfony core team to be able to reproduce this issue directly.

Thank you.

**Comment by jerryladoo** [comment]
@xabbuh @javiereguiluz @Simperfit Sorry guys for such late reply, I am quite busy recently as our own product is about releasing new version. I don't have too much time to create a small project to reproduce this issue at moment, but I will find another time to do it. 

Once I create, I will let you guys know. Sorry about that

**Comment by jerryladoo** [comment]
@xabbuh @javiereguiluz @Simperfit hey guys, I have created [a small project](https://github.com/jerryladoo/bug_19149) in github for reproducing this bug. Hopefully it can help.

**Comment by xabbuh** [comment]
@jerryladoo Thank you for the patience and for putting your effort into creating the example project. I just spotted that you are comparing the processed configuration in your container extension with the unprocessed configurations (note that this is an array of configurations) in your compiler pass. If you had dumped the `$configs` array in your extension you would have observed the same results in both the extension and the compiler pass. So this is the expected behaviour and thus I am closing here.

Though may I ask for what use case your are trying to deal with the configuration in your compiler pass?

**Comment by jerryladoo** [comment]
@xabbuh thanks for explanation. Then you are saying that it is an expected behaviour that the default value assigned in configuration.php should be ignored in compiler pass as the config returned by `$container->getExtensionConfig()` is unprocessed, is my understanding correct?

May then I ask how can I get `$config` in compiler pass without using `$container->getExtensionConfig()`

What I are trying to achieve is to automatically wire up config settings with tagged services based on default value given. For example, if `ZipArchive` is set to `use` in config as default, then `ZipArchiver` should be used, if `tar` is given, then `TarArchiver` should then be used, and so forth, plus also giving `working_dir` to the service.

However, compiler pass is only place I am aware of to have those tagged services grouped by defined attribute (unless there is other way to do so ), so I tried to get the config settings with `getExtensionConfig`, which is the only method I knew to fetch config, for this purpose.

**Comment by xabbuh** [comment]
@jerryladoo Yes, this is totally expected. You could in theory process the configuration again inside the compiler pass. But a better approach is to set some container parameters inside your DI extension that you can then refer to in your compiler pass.

**Comment by jerryladoo** [comment]
@xabbuh  thanks for that, but I still have a question. From execution order perspective, the extension where config gets processed will be called first and then compiler pass. Besides, from the name of method `getExtensionConfig`, it tells me that I should expect to have identical config in compiler pass, in fact it is not. Instead the config in compiler pass seems reverting back to its original state, which only contains those values that have be hard-coded in `config.yml` but not those assigned as default value in `configuration.php`. May I ask why there is such an inconsistent behaviour between extension and compiler pass about processing config?

**Comment by xabbuh** [comment]
In fact there is no difference between a compiler pass and the extension class. Both only no that there is some kind of config provided by the user. The processed configuration isn't stored anywhere. Actually, it even is not a must to process the configuration at all (though there's limited use without it).

**Comment by stof** [comment]
@jerryladoo ``getExtensionConfig`` returns you exactly the same config that the one you receive in your extension (except that the extension receives it as argument as the call to ``getExtensionConfig`` is done before calling the extension itself).

**Comment by jerryladoo** [comment]
@stof, the fact is I am getting a different result between what `getExtensionConfig` returns and the one in `Extension`, one with default value attached, one doesn't.

**Comment by xabbuh** [comment]
@jerryladoo As said you are comparing the config that you passed to the process method. You need to compare the raw config passed to the Extension class with the value you can access in the compiler pass. There is no difference then.

**Comment by jerryladoo** [comment]
If I output `$configs` before `$config = $this->processConfiguration($configuration, $configs);`, I saw the same result as what `getExtensionConfig` returns in compilerPass. I understood this part(if this is what you guys kept saying as 'raw config'). And I took @xabbuh 's suggestion, set the config into `container` after `$config = $this->processConfiguration($configuration, $configs)` like so:
```
...
$container->setParameter("config", $config);
...
```
then picking it up in compiler pass, I will always get a right config because of the order of execution, extension will always get executed first then compiler pass, unless this order will change.

My last question above is my curious. Probably putting in this way will eliminate confusion, why compiler pass don't want to give a processed config, even an optional getter?
